File: src/day09/main.exs
defmodule Day09 do
  def sign(x) do
    cond do
      x > 0 -> 1
      x == 0 -> 0
      x < 0 -> -1
    end
  end

  def follow(path) do
    Stream.transform(path, {0, 0}, fn {hx, hy}, {tx, ty} ->
      if hx - 1 <= tx && tx <= hx + 1 && hy - 1 <= ty && ty <= hy + 1 do
        {[], {tx, ty}}
      else
        p = {tx + sign(hx - tx), ty + sign(hy - ty)}
        {[p], p}
      end
    end)
    |> Enum.to_list()
  end
end

path0 =
  File.read!("puzzle_input/day09")
  |> String.trim_trailing()
  |> String.split(~r/\R+/, trim: True)
  |> Enum.flat_map(fn s ->
    [dir, dist] = String.split(s)
    {dist, _} = Integer.parse(dist)

    step =
      case dir do
        "U" -> {0, 1}
        "D" -> {0, -1}
        "L" -> {-1, 0}
        "R" -> {1, 0}
      end

    Stream.cycle([step])
    |> Stream.take(dist)
  end)
  |> Stream.transform({0, 0}, fn {dx, dy}, {x, y} ->
    p = {x + dx, y + dy}
    {[p], p}
  end)

path1 = Day09.follow(path0)

p1 = Enum.count(Stream.uniq([{0, 0} | path1]))
IO.puts("part 1: #{p1}")

path9 =
  Enum.reduce(2..9, path1, fn _, path ->
    Day09.follow(path)
  end)

p2 = Enum.count(Stream.uniq([{0, 0} | path9]))
IO.puts("part 2: #{p2}")

File: src/day09/main.hs
import Data.List (foldl')
import qualified Data.Set as Set

steps :: String -> [(Int, Int)]
steps s = replicate dist' direction'
  where
    ws = words s
    (direction : dist : _) = ws
    dist' = read dist
    direction' = case direction of
      "U" -> (0, 1)
      "D" -> (0, -1)
      "L" -> (-1, 0)
      "R" -> (1, 0)
      _ -> error "bad direction"

unique :: Ord a => [a] -> [a]
unique values = Set.toList $ Set.fromList values

sign :: (Num a, Num p, Ord a) => a -> p
sign 0 = 0
sign x
  | x > 0 = 1
  | otherwise = -1

follow :: [(Int, Int)] -> [(Int, Int)]
follow path = let (path', _, _) = foldl' followStep ([(0, 0)], 0, 0) path in path'

followStep :: ([(Int, Int)], Int, Int) -> (Int, Int) -> ([(Int, Int)], Int, Int)
followStep (path, tx, ty) (hx, hy) =
  let (dx, dy) = (hx - tx, hy - ty)
   in if (-1 <= dx) && (dx <= 1) && (-1 <= dy) && (dy <= 1)
        then (path, tx, ty)
        else
          let tx' = tx + sign dx
              ty' = ty + sign dy
           in (path ++ [(tx', ty')], tx', ty')

main :: IO ()
main = do
  input <- readFile "puzzle_input/day09"
  let path1 = follow $ reverse $ foldl' takeStep [] $ concatMap steps $ lines input
  putStrLn $ (++) "part 1: " $ show $ length $ unique path1
  let path9 = iterate follow path1 !! 8
  putStrLn $ (++) "part 2: " $ show $ length $ unique path9
  where
    takeStep :: [(Int, Int)] -> (Int, Int) -> [(Int, Int)]
    takeStep ((x, y) : rest) (dx, dy) = (x + dx, y + dy) : (x, y) : rest
    takeStep [] (dx, dy) = [(dx, dy), (0, 0)]